File: android/app/src/main/java/dev/nytweetdeck/android/model/VideoQuality.kt
package dev.nytweetdeck.android.model

fun selectVideoUrl(media: Media, quality: VideoQuality): String? {
    val variants = media.variants
        .filter { it.url.isNotBlank() }
        .sortedBy { it.bitrate ?: 0L }
    if (variants.isEmpty()) return media.url
    val index = when (quality) {
        VideoQuality.AUTO,
        VideoQuality.HIGH,
        -> variants.lastIndex
        VideoQuality.MEDIUM -> (variants.lastIndex / 2)
        VideoQuality.LOW -> 0
    }
    return variants.getOrNull(index)?.url ?: media.url
}

File: frontend/src/model/video-quality.ts
import type { VideoQuality } from "./layout";
import type { VideoVariant } from "./timeline";

export function selectVideoSource(
  fallbackSource: string,
  variants: readonly VideoVariant[] | undefined,
  quality: VideoQuality,
): string {
  return selectVideoSources(fallbackSource, variants, quality)[0] ?? fallbackSource;
}

export function selectVideoSources(
  fallbackSource: string,
  variants: readonly VideoVariant[] | undefined,
  quality: VideoQuality,
): string[] {
  const candidates = (variants ?? [])
    .filter((variant) => variant.url.trim().length > 0)
    .slice()
    .sort((left, right) => (left.bitrate ?? 0) - (right.bitrate ?? 0));
  if (candidates.length === 0) return fallbackSource.trim().length > 0 ? [fallbackSource] : [];

  const index =
    quality === "low"
      ? 0
      : quality === "medium"
        ? Math.floor((candidates.length - 1) / 2)
        : candidates.length - 1;
  const preferred = candidates[index]?.url ?? fallbackSource;
  if (quality !== "auto") return [preferred];
  return [
    ...new Set(
      candidates
        .slice()
        .reverse()
        .map((candidate) => candidate.url),
    ),
  ];
}
